theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,265 @@
+#section support_code
+
+typedef struct ctc_context {
+    struct ctcOptions options;
+    void * workspace;
+    int * input_lengths;
+    int * flat_labels;
+    int * label_lengths;
+} ctc_context_t;
+
+void ctc_context_init(ctc_context_t * context)
+{
+    struct ctcOptions * options = &(context->options);
+    memset(options, 0, sizeof(struct ctcOptions));
+    options->loc = CTC_CPU;
+    options->num_threads = 1;
+
+    context->workspace = NULL;
+    context->input_lengths = NULL;
+    context->flat_labels = NULL;
+    context->label_lengths = NULL;
+}
+
+void ctc_context_destroy(ctc_context_t * context)
+{
+    if ( NULL != context->workspace )
+        free( context->workspace );
+
+    if ( NULL != context->input_lengths )
+        free( context->input_lengths );
+
+    if ( NULL != context->flat_labels )
+        free( context->flat_labels );
+
+    if ( NULL != context->label_lengths )
+        free( context->label_lengths );
+}
+
+int ctc_check_result(ctcStatus_t retcode, const char * msg)
+{
+    if( CTC_STATUS_SUCCESS != retcode )
+    {
+        // Get error message from underlying library
+        const char * ctc_msg = ctcGetStatusString( retcode );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "%s | CTC library error message: %s",
+                      msg,
+                      ctc_msg );
+        return 1;
+    }
+    return 0;
+}
+
+void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr,
+    int ** input_lengths )
+{
+    int num_elements = PyArray_DIMS( input_lengths_arr )[0];
+
+    *input_lengths = (int *) malloc( num_elements * sizeof(int) );
+
+    if ( NULL == (*input_lengths) )
+        return;
+
+    for( int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
+    {
+        (*input_lengths)[elem_idx] = *( (int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
+    }
+}
+
+void create_flat_labels( PyArrayObject * label_matrix, int ** flat_labels,
+    int ** label_lengths )
+{
+    int rows = PyArray_DIMS( label_matrix )[0];
+    int cols = PyArray_DIMS( label_matrix )[1];
+
+    *flat_labels = (int *) malloc( rows * cols * sizeof(int) );
+    if ( NULL == (*flat_labels) )
+        return;
+
+    *label_lengths = (int *) malloc( rows * sizeof(int) );
+    if ( NULL == (*label_lengths) )
+    {
+        free( *flat_labels );
+        *flat_labels = NULL;
+        return;
+    }
+
+    int label_index = 0;
+    for( int row_idx = 0; row_idx < rows; ++row_idx )
+    {
+        int label_length = 0;
+        for( int col_idx = 0; col_idx < cols; ++col_idx )
+        {
+            int label = *( (int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
```

Comment:
label_matrix pointer should be casted to npy_int32 instead of int, as label_matrix should be an int32 tensor (checked in make_node()). It could then be dereferenced and copied to an int without any loss, so the rest of code for this function should be still ok.

```C
int label = *( (npy_int32 *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
```